src/sql/src/plan/expr.rs

```diff
@@ -401,6 +401,30 @@ impl AbstractExpr for CoercibleScalarExpr {
     ) -> Self::Type {
         match self {
             CoercibleScalarExpr::Coerced(expr) => Some(expr.typ(outers, inner, params)),
+            CoercibleScalarExpr::LiteralRecord(exprs) => {
+                let mut field_counter = 1;
+                let fields: Vec<_> = exprs
+                    .iter()
+                    .filter_map(|expr| expr.typ(outers, inner, params))
+                    .map(|typ| {
+                        let field_name = ColumnName::from(format!("f{}", field_counter));
+                        field_counter += 1;
+                        (field_name, typ.nullable(false))
+                    })
+                    .collect();
+                if fields.len() == exprs.len() {
+                    Some(
+                        ScalarType::Record {
+                            fields,
+                            custom_oid: None,
+                            custom_name: None,
+                        }
+                        .nullable(true),
+                    )
+                } else {
+                    None
+                }
+            }
```

Comment:
I've paged out a lot of the context, but this looks like https://github.com/MaterializeInc/materialize/pull/10868 to me, which didn't seem quite right to me at the time. Wrote down details here! https://github.com/MaterializeInc/materialize/issues/10934#issue-1153535478

Reply:
ty for the pointer; will take a look at that and see if there's a way around this.